ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,616 @@
+            entry.getValue() <= blobBalanceThreshold);
+      }
+    }
+  }
+
+  /**
+   * Create an {@link OperationChain} from a queue of {@link OperationType}s.  Start the operation chain asynchronously.
+   * @param blobSize the size of the blob generated for put operations.
+   * @param chainId a numeric identifying the operation chain.
+   * @param operations the queue of operations to perform in the chain
+   * @return an {@link OperationChain} object describing the started chain.
+   */
+  OperationChain startOperationChain(int blobSize, int chainId, Queue<OperationType> operations) {
+    byte[] userMetadata = new byte[1000];
+    byte[] data = new byte[blobSize];
+    new Random().nextBytes(userMetadata);
+    new Random().nextBytes(data);
+    BlobProperties properties = new BlobProperties(blobSize, "serviceid1");
+    OperationChain opChain = new OperationChain(chainId, properties, userMetadata, data, operations);
+    continueChain(opChain);
+    return opChain;
+  }
+
+  /**
+   * Generate the properties needed by the router and coordinator.  NOTE: Properties for SSL interaction need
+   * to be added manually.
+   * @param routerDatacenter the datacenter name where the router will be running.
+   * @return a {@link Properties} object with the properties needed to instantiate the router/coordinator.
+   */
+  static Properties getRouterProperties(String routerDatacenter) {
+    Properties properties = new Properties();
+    properties.setProperty("router.hostname", "localhost");
+    properties.setProperty("router.datacenter.name", routerDatacenter);
+    properties.setProperty("router.connection.checkout.timeout.ms", "5000");
+    properties.setProperty("router.request.timeout.ms", "10000");
+    properties.setProperty("router.max.put.chunk.size.bytes", Integer.toString(CHUNK_SIZE));
+    properties.setProperty("router.put.success.target", "1");
+    properties.setProperty("coordinator.hostname", "localhost");
+    properties.setProperty("coordinator.datacenter.name", routerDatacenter);
+    return properties;
+  }
+
+  /**
+   * Check for blob ID validity.
+   * @param blobId the blobId
+   * @param operationName a name for the operation being checked
+   */
+  private static void checkBlobId(String blobId, String operationName) {
+    Assert.assertNotNull("Null blobId for operation: " + operationName, blobId);
+  }
+
+  /**
+   * Check that {@code blobInfo} matches {@code opChain.blobInfo}.
+   * @param blobInfo the {@link BlobInfo} to check
+   * @param opChain the {@link OperationChain} structure to compare against
+   * @param operationName a name for the operation being checked
+   */
+  private static void checkBlobInfo(BlobInfo blobInfo, OperationChain opChain, String operationName) {
+    Assert.assertNotNull("Null blobInfo for operation: " + operationName, blobInfo);
+    Assert.assertEquals("Blob size in info does not match expected for operation: " + operationName,
```

Comment:
`RouterTestHelpers.java` has a method called `haveEquivalentFields()` for comparing `BlobProperties`, do you want to use that? (just fyi, we decided not to override `equals()` for `BlobProperties` itself due to the fact that it would break equality when comparison happens between `BlobProperties` and its derived class `SystemMetadata`).